text/3502-cargo-script.md

```diff
@@ -0,0 +1,1202 @@
+  - Based on feedback, we might add `cargo-<edition>-edition` proxies to put in `#!` as a shorthand
+  - Based on feedback, we can switch to "edition is required as of <future> edition"
+
+Disallowed manifest fields:
+- `[workspace]`, `[lib]`, `[[bin]]`, `[[example]]`, `[[test]]`, `[[bench]]`
+- `package.workspace`, `package.build`, `package.links`, `package.publish`, `package.autobins`, `package.autoexamples`, `package.autotests`, `package.autobenches`
+
+Single-file packages maintain an out-of-line target directory by default.
+This is `$CARGO_HOME/target/<hash-of-path>`.
+
+A single-file package is accepted by cargo commands as a `--manifest-path`
+- Files are considered to have embedded manifest if they end with `.rs` or they lack and extension and are of type file
+- This allows running `cargo test --manifest-path single.rs`
+- `cargo add` and `cargo remove` may not support editing embedded manifests initially
+- Path-dependencies may not refer to single-file packages at this time (they don't have a `lib` target anyways)
+
+Single-file packages will not be accepted as `path` or `git` dependencies.
+
+The lockfile for single-file packages will be placed in `CARGO_TARGET_DIR`.  In
+the future, when workspaces are supported, that will allow a user to have a
+persistent lockfile.
+We may also allow customizing the non-workspace lockfile location in the [future](#future-possibilities).
+
+## `cargo <file>.rs`
+
+`cargo` is intended for putting in the `#!` for single-file packages:
+```rust
+#!/usr/bin/env cargo
+
+fn main() {
+    println!("Hello world");
+}
+```
+- Like with `cargo install`, `.cargo/config.toml` will be read based on the
+  scripts location rather than the current-dir.
+  - And like `cargo install`, the current-dir rustup-toolchain is respected
+- `--release` is not passed in because the primary use case is for exploratory
+  programming, so the emphasis will be on build-time performance and debugging,
+  rather than runtime performance
+
+Most other flags and behavior will be similar to `cargo run`.
+
+The precedence for `cargo foo` will change from:
+1. built-in commands
+2. user aliases
+3. third-party commands
+
+to:
+1. built-in command xor manifest
+2. user aliases
+3. third-party commands
+
+To allow the xor, we enforce that
+- manifests must be passed in as `Cargo.toml`, with a `.rs` extension, or have a `/` in them
+  - So `./build` can be used to run a script name `build` rather than the `cargo build` command
+- no built-in command may look like an accepted manifest
+
+When piping `cargo <file>.rs`, `--quiet` will be assumed.
+Further work may be done to refine the output in interactive mode.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The implicit content of the manifest will be unclear for users.
+We can patch over this as best we can in documentation but the result won't be
+ideal.
+A user can workaround this with `cargo metadata --manifest-path <file>.rs`.
```

Comment:
Idea for discoverability: Make `cargo --verbose <file>.rs` print the implicit manifest along with running the script. "Explain the steps we are taking to interpret the input" is in line with what people expect from `--verbose`.

Reply:
The challenge with `--verbose` (or log levels) is everyone (and every situation) has a different idea of what extra output is needed.

I feel like `cargo metadata` and `cargo read-manifest` are likely sufficient.